File: app/src/main/java/br/com/disapps/homepet/ui/common/AppFragment.java
package br.com.disapps.homepet.ui.common;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.DrawableRes;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.hannesdorfmann.mosby3.mvp.MvpFragment;
import com.hannesdorfmann.mosby3.mvp.MvpPresenter;

import java.net.HttpURLConnection;

import br.com.disapps.homepet.util.rx.RxHttpError;
import butterknife.ButterKnife;
import butterknife.Unbinder;

/**
 * Created by diefferson.santos on 23/08/17.
 */

public abstract class AppFragment<V extends AppView, P extends MvpPresenter<V>> extends MvpFragment<V, P> implements IErrorHandlerView, AppView {

    private IAppActivityListener iAppActivityListener;
    private Unbinder unbinder;

    @Override
    public boolean hasInternetConnection() {
        return getAppActivityListener().hasInternetConnection();
    }
    public void showLoading() {
//        if (loadingFragment == null) {
//            loadingFragment = LoadingFragment.newInstance();
//        }
//        loadingFragment.show(getFragmentManager(), LoadingFragment.class.getSimpleName());
    }

    public void dismissLoading() {
//        loadingFragment.dismiss();
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        unbinder = ButterKnife.bind(this, view);

    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(getFragmentLayout(), container, false);
    }

    @Override
    public void onDestroyView() {
        unbinder.unbind();
        super.onDestroyView();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        iAppActivityListener = (IAppActivityListener) context;
    }

    @Override
    public void error(RxHttpError error) {
        switch (error.errorCode) {
            case RxHttpError.SOCKETTIMEOUT_CODE:
            case RxHttpError.UNKNOWNHOST_CODE:
//                setMessageContent(true, R.string.err_unknownhost, getSectionIcPlaceholder());
                break;
            case RxHttpError.NO_CONNECTIVITY_CODE:
//                setMessageContent(true, R.string.err_connectivity, getSectionIcPlaceholder());
                break;
            case HttpURLConnection.HTTP_UNAUTHORIZED:
//                CustomApplication.getInstance().doLogout( getContext() );
                break;
            case 422:
//                setMessageContent(true, error.detail, getSectionIcPlaceholder());
                break;
            default:
//                setMessageContent(true, R.string.err_default, getSectionIcPlaceholder());
                break;
        }
    }

    @Override
    public void error(Throwable e) {
        e.printStackTrace();
//        setMessageContent(true, R.string.err_default, getSectionIcPlaceholder());
    }

    protected void setMessageContent(boolean isVisible, @StringRes int resMessage, @DrawableRes int resIcon) {
//        setMessageContent(isVisible, getContext().getString(resMessage), resIcon);
    }


    /**
     * Tag para identificacao do Fragmento na pilha do FragmentManager
     *
     * @return Tag que identifica o fragmento
     */
    public abstract String getFragmentTag();

    public abstract int getFragmentLayout();

    protected IAppActivityListener getAppActivityListener() {
        return iAppActivityListener;
    }
}

File: app/src/main/java/br/com/disapps/homepet/ui/common/AppFragment.kt
package br.com.disapps.homepet.ui.common

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.hannesdorfmann.mosby3.mvp.MvpFragment
import com.hannesdorfmann.mosby3.mvp.MvpPresenter
import java.net.HttpURLConnection
import br.com.disapps.homepet.util.rx.RxHttpError

/**
 * Created by diefferson.santos on 23/08/17.
 */

abstract class AppFragment<V : AppView, P : MvpPresenter<V>> : MvpFragment<V, P>(), IErrorHandlerView, AppView {

    protected var appActivityListener: IAppActivityListener? = null
        private set

    private var loadingView: View? = null

    override fun hasInternetConnection(): Boolean {
        return appActivityListener!!.hasInternetConnection()
    }

    fun setupLoadingFragment(loadingView: View) {
        this.loadingView = loadingView
    }

    override fun showLoading(cancelable: Boolean) {
        if (loadingView != null) {
            loadingView!!.visibility = View.VISIBLE
        }
    }

    override fun dismissLoading() {
        if (loadingView != null) {
            loadingView!!.visibility = View.GONE
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(fragmentLayout, container, false)
    }
    
    override fun onDestroyView() {
        super.onDestroyView()
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        appActivityListener = context as IAppActivityListener?
    }

    override fun error(error: RxHttpError) {
        when (error.errorCode) {
            RxHttpError.SOCKETTIMEOUT_CODE, RxHttpError.UNKNOWNHOST_CODE -> setMessageContent(error.errorMessage)
            RxHttpError.NO_CONNECTIVITY_CODE -> setMessageContent(error.errorMessage)
            HttpURLConnection.HTTP_UNAUTHORIZED -> setMessageContent("Usuário ou senha invalidos")
            HttpURLConnection.HTTP_CONFLICT -> setMessageContent("Usuário já cadastrado.")
            422 -> setMessageContent(error.errorMessage)
            else -> setMessageContent(error.errorMessage)
        }
    }

    override fun error(e: Throwable) {
        e.printStackTrace()
        setMessageContent(e.message!!)
    }

    protected fun setMessageContent(message: String) {
        Toast.makeText(context, message, Toast.LENGTH_LONG).show()
    }

    abstract val fragmentTag: String

    abstract val fragmentLayout: Int

}
